File: app/src/main/java/com/arnaudpiroelle/manga/core/utils/HttpUtils.java
package com.arnaudpiroelle.manga.core.utils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;
import java.nio.channels.WritableByteChannel;

import retrofit.client.Response;

public class HttpUtils {
    public static String convertFrom(Response response) {
        BufferedReader reader;
        StringBuilder sb = new StringBuilder();
        try {
            reader = new BufferedReader(new InputStreamReader(response.getBody().in()));

            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return sb.toString();
    }

    public static InputStream readFrom(Response response) {
        try {
            return response.getBody().in();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return null;
    }

    public static void writeFile(InputStream inputStream, File pageFile) throws IOException {
        FileOutputStream outputStream = new FileOutputStream(pageFile);

        ReadableByteChannel inputChannel = Channels.newChannel(inputStream);
        WritableByteChannel outputChannel = Channels.newChannel(outputStream);

        fastChannelCopy(inputChannel, outputChannel);

        inputChannel.close();
        outputChannel.close();
    }

    private static void fastChannelCopy(final ReadableByteChannel src, final WritableByteChannel dest) throws IOException {
        final ByteBuffer buffer = ByteBuffer.allocateDirect(16 * 1024);
        while (src.read(buffer) != -1) {
            buffer.flip();
            dest.write(buffer);
            buffer.compact();
        }

        buffer.flip();

        while (buffer.hasRemaining()) {
            dest.write(buffer);
        }
    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/core/utils/HttpUtils.kt
package com.arnaudpiroelle.manga.core.utils

import retrofit.client.Response
import java.io.*
import java.nio.ByteBuffer
import java.nio.channels.Channels
import java.nio.channels.ReadableByteChannel
import java.nio.channels.WritableByteChannel

object HttpUtils {
    fun convertFrom(response: Response): String {
        val reader: BufferedReader
        val sb = StringBuilder()
        try {
            reader = BufferedReader(InputStreamReader(response.body.`in`()))

            var line: String? = reader.readLine()

            while (line != null) {
                sb.append(line)
                line = reader.readLine()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return sb.toString()
    }

    fun readFrom(response: Response): InputStream {
        return response.body.`in`()
    }

    @Throws(IOException::class)
    fun writeFile(inputStream: InputStream, pageFile: File) {
        val outputStream = FileOutputStream(pageFile)

        val inputChannel = Channels.newChannel(inputStream)
        val outputChannel = Channels.newChannel(outputStream)

        fastChannelCopy(inputChannel, outputChannel)

        inputChannel.close()
        outputChannel.close()
    }

    @Throws(IOException::class)
    private fun fastChannelCopy(src: ReadableByteChannel, dest: WritableByteChannel) {
        val buffer = ByteBuffer.allocateDirect(16 * 1024)
        while (src.read(buffer) != -1) {
            buffer.flip()
            dest.write(buffer)
            buffer.compact()
        }

        buffer.flip()

        while (buffer.hasRemaining()) {
            dest.write(buffer)
        }
    }
}
